include/deal.II/lac/cuda_vector.h

```diff
@@ -0,0 +1,277 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__cuda_vector_h
+#define dealii__cuda_vector_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/lac/vector_space_vector.h>
+
+
+#ifdef DEAL_II_WITH_CUDA
+
+DEAL_II_NAMESPACE_OPEN
+
+class CommunicationPatternBase;
+class IndexSet;
+template <typename Number> class ReadWriteVector;
+
+namespace LinearAlgebra
+{
+  namespace CUDAWrappers
+  {
+    /**
+     * This class implement vector using CUDA for use on Nvidia GPUs. This
+     * class is derived from the LinearAlgebra::VectorSpaceVector class.
+     *
+     * @ingroup CUDAWrappers
+     * @ingroup Vectors
+     * @author Karl Ljungkvist, Bruno Turcksin, 2016
+     */
+    template <typename Number>
+    class Vector: public VectorSpaceVector<Number>
+    {
+    public:
+      typedef typename VectorSpaceVector<Number>::size_type  size_type;
+      typedef typename VectorSpaceVector<Number>::real_type  real_type;
+
+      /**
+       * Constructor. Create a vector of dimension zero.
+       */
+      Vector();
+
+      /**
+       * Copy constructor.
+       */
+      Vector(const Vector<Number> &V);
+
+      /**
+       * Constructor. Set dimension to @p n and initialize all elements with
+       * zero.
+       *
+       * The constructor is made explicit to avoid accident like this:
+       * <tt>v=0;</tt>. Presumably, the user wants to set every elements of
+       * the vector to zero, but instead, what happens is this call:
+       * <tt>v=Vector@<Number@>(0);</tt>, i.e. the vector is replaced by one
+       * of length zero.
+       */
+      explicit Vector(const size_type n);
+
+      /**
+       * Destructor.
+       */
+      ~Vector();
+
+      /**
+       * Reinit functionality. The flag <tt>omit_zeroing_entries</tt>
+       * determines wheter the vector should be filled with zero (false) or
+       * left untouched (true).
+       */
+      void reinit(const size_type n,
+                  const bool      omit_zeroing_entries = false);
+
+      /**
+       * Import all the element from the input vector @p V.
+       * VectorOperation::values @p operation is used to decide if the
+       * elements int @p V should be added to the current vector or replace
+       * the current elements. The last parameter is not used. It is only used
+       * for distributed vectors. This is the function that should be used to
+       * copy a vector to the GPU.
+       */
+      virtual void import(const ReadWriteVector<Number> &V,
+                          VectorOperation::values operation,
+                          std_cxx11::shared_ptr<const CommunicationPatternBase> communication_pattern =
+                            std_cxx11::shared_ptr<const CommunicationPatternBase> ()) override;
```

Comment:
We have decided that CUDA will require C++11. You can't use an old compiler to compile cuda anyway.